feature: RFC3021 (/31) and /32 support added to snmp-scan.php

iviemeister opened the pull request:
#### Please note

> Please read this information carefully. You can run `./scripts/pre-commit.php` to check your code before submitting.

- [x] Have you signed the [Contributors agreement](http://docs.librenms.org/General/Contributing/)
- [x] Have you followed our [code guidelines?](http://docs.librenms.org/Developing/Code-Guidelines/)

Support RFC3021 - Add /31 and /32 subnet support to snmp-scan.php, add force-scan options for network and broadcast addresses, improve debugging messages.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/librenms/librenms/inspections/5b105850-8acc-4944-87d0-be05f0af0e10) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/librenms/librenms/inspections/5b105850-8acc-4944-87d0-be05f0af0e10) completed: **4 new issues, 1 updated code elements**

laf commented on snmp-scan.php:
```diff
@@ -46,62 +46,131 @@
 function perform_snmp_scan($net)
 {
-    global $stats, $config, $debug, $vdebug;
+    global $stats, $config, $debug, $vdebug, $force_network, $force_broadcast;
```
You should probably pass these in to the function rather than using global.

laf commented on snmp-scan.php:
```diff
@@ -46,62 +46,131 @@
     $start = ip2long($net->network);
     $end   = ip2long($net->broadcast)-1;
+
+    if ($force_network == true) { //Force-scan network address
```
Needs to be ===

laf commented on snmp-scan.php:
```diff
@@ -46,62 +46,131 @@
     $end   = ip2long($net->broadcast)-1;
+
+    if ($force_network == true) { //Force-scan network address
+        if ($debug == true) {
```
Please use d_echo() function instead of `if ($debug) { echo ''; }`

laf commented on snmp-scan.php:
```diff
@@ -46,62 +46,131 @@
+        $start = $start-1;
+    }
+
+    if ($force_broadcast == true) { //Force-scan broadcast address
```
All == true should be === to be explicit.

laf commented on snmp-scan.php:
```diff
@@ -46,62 +46,131 @@
+    }
+
+    if ($force_broadcast == true) { //Force-scan broadcast address
+        if ($debug == true) {
```
Same comment here for d_echo()

laf commented on snmp-scan.php:
```diff
@@ -46,62 +46,131 @@
+        $end = $end+1;
+    }
+
+    if ($net->bitmask == 31) { //Handle RFC3021 /31 prefixes
```
Can use === here as well.

laf commented on snmp-scan.php:
```diff
@@ -46,62 +46,131 @@
+        $end   = $start+1;
+    }
+
+    if ($debug == true) {
```
Can remove the if $debug check and just wrap the echo's in d_echo instead.

laf commented on snmp-scan.php:
```diff
@@ -46,62 +46,131 @@
         $stats['count']++;
         $host = long2ip($start);
+
+        if ($vdebug == true) {
```
d_echo()

laf replied:
`===`

laf commented on snmp-scan.php:
```diff
@@ -46,62 +46,131 @@
+
         if (match_network($config['autodiscovery']['nets-exclude'], $host)) {
-            echo '|';
+            if ($vdebug == true) {
```
d_echo()

laf replied:
===

laf commented on snmp-scan.php:
```diff
@@ -46,62 +46,131 @@
+    if ($debug == true) {
+        if ($net->bitmask == 31) {
+            echo "RFC3021 network, hosts ".long2ip($start+1)." and ".long2ip($end).PHP_EOL.PHP_EOL;
+        } elseif ($net->bitmask == 32) {
```
`===`

laf commented on snmp-scan.php:
```diff
@@ -46,62 +46,131 @@
+    }
+
+    if ($debug == true) {
+        if ($net->bitmask == 31) {
```
`===`

laf commented on snmp-scan.php:
```diff
@@ -46,62 +46,131 @@
         $test = isPingable($host);
         if ($test['result'] === false) {
-            echo '.';
+            if ($vdebug == true) {
```
Remove if ($vdebug) and wrap echo in d_echo()

laf commented on snmp-scan.php:
```diff
@@ -46,62 +46,131 @@
         if (ip_exists($host)) {
             $stats['known']++;
-            echo '*';
+            if ($vdebug == true) {
```
Removed $vdebug and use d_echo()

laf commented:
Thanks for the contribution. I've put some inline comments in place but stopped near the end, those comments apply all the way through though.

`===` should be used for strict checking.

`d_echo()` is available to use instead of checking if debug is set then echoing out.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/librenms/librenms/inspections/7f156cd4-1a67-4a3e-8659-aa221d77423f) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/librenms/librenms/inspections/7f156cd4-1a67-4a3e-8659-aa221d77423f) completed: **1 updated code elements**

iviemeister commented:
Switched all == to ===, switched all "if ($debug..." to d_echo().

Left "if ($vdebug..." in scan loop, as those actually change output, not just add, and only in verbose debug mode.

laf commented:
This has been merged now but $vdebug isn't for showing additional output (it's badly named really). It was put in to show the masked information we have in debug which hides sensitive info so really this still should be in d_echo()

f0o commented:
Darn, i can do a follow-up patch
